### PR TITLE
Add 'jacobi' sample to Tools/ApplicationDebugger.

### PR DESCRIPTION
**Pull request by nsaiapova**
Signed-off-by: Natalia Saiapova <natalia.saiapova@intel.com>

# Adding a New sample
##Introduction

- Submission of new Sample `jacobi` for Application Debugger. It solves a hardcoded linear system Ax=b with Jacobi iteration. We intentionally introduced several bugs. It helps to try out the debugger to find real issues.

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): @barisaktemur 
- [X] If you dependencies/binaries, inform the oneAPI Code Samples Project Manager: @JoeOster **no dependencies added**

Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [X] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see @JoeOster for assistance)
- [X] Compile using the following compiler flags and fix any warnings, the flags are: "/Wall -Wformat-security -Werror=format-security"
- [x] Protex (see @JoeOster for assistance)
- [ ] Bandit Scans (Python only) **N/A**
- [x] IPLDT Scan (see @JoeOster for assistance)
- [x] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: @pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [x] Tested using Dev Cloud when applicable

### CODE REVIEW

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
+# `jacobi` Sample
```
Comment (JoeOster) [inline Tools/ApplicationDebugger/jacobi/README.md:1]: I made several minor edits, please take a look and make sure you are in agreement with changes

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    $ cmake ..
+    $ make
+    ```
+    > Note: The cmake configuration enforces the `Debug` build type.
+
+2.  Run the bugged program:
+    ```
+    $ ./jacobi-bugged <device>
+    ```
+    > Note: `<device>` is the type of the device type to offload the kernel.
```
Comment (JoeOster) [inline Tools/ApplicationDebugger/jacobi/README.md:129]: Sentence needs restructuring

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+The include folder is located at
+`%ONEAPI_ROOT%\dev-utilities\latest\include` on your development
+system.
+
+
+### Running Samples In DevCloud
+
+If running a sample in the Intel DevCloud, remember that you must
+specify the compute node (CPU, GPU, FPGA) and whether to run in
```
Comment (JoeOster) [inline Tools/ApplicationDebugger/jacobi/README.md:99]: I was told that the interactive mode is pretty hidden at the moment and that the recommendation is to just qsub (batch mode). The big problem with interactive is that it ties up resources for 6 hrs. Please verify the links and make sure the info is there that's needed

**Review by JoeOster — CHANGES_REQUESTED** [review]

File: Tools/ApplicationDebugger/jacobi/sample.json [hunk 4]
```diff
@@ -0,0 +1,174 @@
+{
+  "guid": "CFAC8EF8-E362-4054-ABE3-A04EEBEBCB46",
+  "name": "Debugger: Jacobi",
+  "categories": ["Toolkit/Get Started", "Toolkit/oneAPI Tools/Application Debugger"],
```
Comment (tomlenth) [inline Tools/ApplicationDebugger/jacobi/sample.json:4]: We added "Debugger" to array transform to ensure it was ordered after vector add in the sample browser/IDE.  we don't want this in the Get Started category either.  so, remove name should be: "name": "Jacobi", and category:   "categories": ["Toolkit/oneAPI Tools/Application Debugger"],

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
+# `jacobi` Sample
+
+The 'jacobi' code sample is a small DPC++ code sample for exercising application debugging using
+Intel&reg; Distribution for GDB\*. It is highly recommended that you go
+through this sample *after* you familiarize yourself with the [array-transform](https://github.com/oneapi-src/oneAPI-samples/tree/bed8a563ace4a31d00a4f1055f65d680fa4fe0a8/Tools/ApplicationDebugger/array-transform)
+sample and the 
+[Get Started Guide](https://software.intel.com/en-us/get-started-with-debugging-dpcpp)
+for the debugger.
+
+This sample contains two versions of the same program. One works correctly,
```
Comment (tomlenth) [inline Tools/ApplicationDebugger/jacobi/README.md:10]: overall this README is great.  one _minor_ suggestion would be to make clear here that by default, both programs are built.  The user does not need to do anything to select one over the other.

**Review by tomlenth — COMMENTED** [review]
minor readme recommendation, some json changes

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 3 above)
Comment (barisaktemur) [reply Tools/ApplicationDebugger/jacobi/README.md:99]: @JoeOster, batch mode would practically be unusable for the debugger exercise.

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+or for the Gen12 family:
+
+```
+$ cmake .. -DDPCPP_COMPILE_TARGET=gen12LP
+```
+
+> *Note:* AoT compilation is particularly helpful in larger
+> applications where compiling with debug information takes
+> considerably longer time.
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/README.md:171]: If we remove the AoT options from the CMake file, we should remove this text, as well.
Is there a compilation time issue with this Jacobi sample?

File: Tools/ApplicationDebugger/jacobi/jacobi-bugged.vcxproj.user [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
+    <LocalDebuggerCommandArguments>cpu</LocalDebuggerCommandArguments>
+    <DebuggerFlavor>WindowsLocalDebugger</DebuggerFlavor>
+    <RemoteDebuggerCommandArguments>gpu</RemoteDebuggerCommandArguments>
+    <RemoteDebuggerEnvironment>CL_CONFIG_USE_NATIVE_DEBUGGER=1</RemoteDebuggerEnvironment>
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/jacobi-bugged.vcxproj.user:7]: Should this be defined for the local env, too?

File: Tools/ApplicationDebugger/jacobi/jacobi-fixed.vcxproj.user [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
+    <LocalDebuggerCommandArguments>cpu</LocalDebuggerCommandArguments>
+    <DebuggerFlavor>WindowsLocalDebugger</DebuggerFlavor>
+    <RemoteDebuggerCommandArguments>gpu</RemoteDebuggerCommandArguments>
+    <LocalDebuggerEnvironment>CL_CONFIG_USE_NATIVE_DEBUGGER=1</LocalDebuggerEnvironment>
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/jacobi-fixed.vcxproj.user:7]: Should this be defined for the remote env, too?

File: Tools/ApplicationDebugger/jacobi/jacobi.filters [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+      <UniqueIdentifier>{93995380-89BD-4b04-88EB-625FBE52EBFB}</UniqueIdentifier>
+      <Extensions>h;hh;hpp;hxx;hm;inl;inc;ipp;xsd</Extensions>
+    </Filter>
+    <Filter Include="Resource Files">
+      <UniqueIdentifier>{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}</UniqueIdentifier>
+      <Extensions>rc;ico;cur;bmp;dlg;rc2;rct;bin;rgs;gif;jpg;jpeg;jpe;resx;tiff;tif;png;wav;mfcribbon-ms</Extensions>
+    </Filter>
+  </ItemGroup>
+  <ItemGroup>
+    <ClCompile Include="src\jacobi.cpp">
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/jacobi.filters:18]: Does this file exist?

File: Tools/ApplicationDebugger/jacobi/src/jacobi-bugged.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        auto acc_b = buffer_b.get_access<access::mode::read>(h);
+        auto acc_x_k = buffer_x_k.get_access<access::mode::read>(h);
+        auto acc_x_k1 = buffer_x_k1.get_access<access::mode::write>(h);
+
+        // kernel-start
+        h.parallel_for(range{n}, [=](id<1> index) {
+          int i = index[0];
+          acc_x_k1[i] = acc_b[i] / 4 - (acc_x_k[i - 2] + acc_x_k[i - 1] +
+                                        acc_x_k[i + 1] + acc_x_k[i + 2]) /
+                                           4;
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/src/jacobi-bugged.cpp:65]: Is indentation off here?
Would it be OK to use local variable to improve readability? E.g.
```cpp
float sum = acc_x_k[i - 2] + acc_x_k[i - 1] + acc_x_k[i + 1] + acc_x_k[i + 2];
acc_x_k1[i] = acc_b[i] / 4 - sum / 4;
```

File: Tools/ApplicationDebugger/jacobi/third-party-programs.txt [hunk 11]
```diff
@@ -0,0 +1,253 @@
+oneAPI Code Samples - Third Party Programs File
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/third-party-programs.txt:1]: Is this file required? Please see my comment in the README.

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+but in the other version, several bugs were injected. You can try to find and fix them 
+using the debugger.
+
+| Optimized for       | Description
+|---------------------|--------------
+| OS                  | Linux Ubuntu 18.04 to 20.04, CentOS* 8, Fedora* 30, SLES 15; Windows* 10
+| Hardware            | Kaby Lake with GEN9 (on GPU) or newer (on CPU)
+| Software            | Intel&reg; oneAPI DPC++/C++ Compiler
+| What you will learn | Find existing bugs in a program using the debugger.
+| Time to complete    | 20 minutes for CPU or FPGA emulator; 30 minutes for GPU
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: Put 1 hour for CPU/FPGA and 2 hours for GPU. @tomlenth do you have a better estimation?

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+
+## Purpose
+
+The `jacobi` sample is a DPC++ application that solves a hardcoded 
+linear system of equations `Ax=b` using the Jacobi iteration. The matrix `A` 
+is `n x n` sparse matrix with diagonals [1 1 4 1 1], and vector `b` has
+a format `[6 7 8 ... 8 7 6]^T`. Thus, the solution vector `x` has a form of `[1 1 ... 1]^T`.
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: I rephrased this section.

File: Tools/ApplicationDebugger/jacobi/jacobi-bugged.vcxproj.user (hunk 7 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/jacobi-bugged.vcxproj.user]: This should be defined only for local environment, thanks!

File: Tools/ApplicationDebugger/jacobi/jacobi-fixed.vcxproj.user (hunk 8 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/jacobi-fixed.vcxproj.user:7]: I would say no, as we use remote for GPU most of the time.

File: Tools/ApplicationDebugger/jacobi/jacobi.filters (hunk 9 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/jacobi.filters]: I removed this item, the project still seem to work fine.

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+In the code, `x^{k+1}` corresponds to the variable `x_k1`, and
+`x^k` corresponds to `x_k`. At the end of each iteration, we update `x_k` with
+the value of `x_k1`.
+
+## License
+
+Code samples are licensed under the MIT license. See
+[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.
+
+Third-party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: Looks like other samples (not only array-transform) have this file in their folders. @JoeOster is this file indeed required?

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+returns an incorrect result. With the debugger, you can put a breakpoint in and 
+outside of the kernel, step through the code, and check the values of local variables. 
+Use these features to find places in `jacobi-bugged` where the program does not 
+behave as it was intended.
+
+The sample is intended for exercising the debugger, not for performance benchmarking.
+
+The debugger supports debugging kernels that run on the CPU, GPU, or
+accelerator devices.  For convenience, the `jacobi`
+code sample provides the ability to select the target device by using a command-line argument of `cpu`, `gpu`, or `accelerator`.
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: Done.

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 5 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: The single "make" builds both targets, in VS the solution build also builds both project, "jacobi-bugged" is set as a startup project.

File: Tools/ApplicationDebugger/jacobi/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+  # Do nothing.  We are already set.
+elseif (DPCPP_COMPILE_TARGET MATCHES cpu)
+  set (CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fsycl-targets=spir64_x86_64-unknown-linux-sycldevice")
+elseif (DPCPP_COMPILE_TARGET MATCHES fpga-emu)
+  set (CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fsycl-targets=spir64_fpga-unknown-linux-sycldevice")
+else ()
+  set (CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fsycl-targets=spir64_gen-unknown-linux-sycldevice")
+  set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Xs \"-device ${DPCPP_COMPILE_TARGET} -internal_options -cl-kernel-debug-enable -options -cl-opt-disable\"")
+endif ()
+message (STATUS "DPCPP_COMPILE_TARGET is '${DPCPP_COMPILE_TARGET}'")
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/CMakeLists.txt]: Ok, I removed the DPCPP_COMPILE_TARGET logic.

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+
+## Building and Running the `jacobi` Program
+
+> Note: if you have not already done so, set up your CLI 
+> environment by sourcing  the setvars script located in 
+> the root of your oneAPI installation. 
+>
+> Linux Sudo: . /opt/intel/oneapi/setvars.sh  
+> Linux User: . ~/intel/oneapi/setvars.sh  
+> Windows: C:\Program Files(x86)\Intel\oneAPI\setvars.bat
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: Changed to normal text.

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+    ```
+    $ cd jacobi
+    $ mkdir build
+    $ cd build
+    $ cmake ..
+    $ make
+    ```
+    > Note: The cmake configuration enforces the `Debug` build type.
+
+2.  Run the bugged program:
```
[the remark this answers is not in the document]
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: Fixed.

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 2 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: I removed the second "type": `<device>` is the type of the device to offload the kernel. Ok with this?

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 6 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: Done, no Jacobi compiles fast enough.

File: Tools/ApplicationDebugger/jacobi/src/jacobi-bugged.cpp (hunk 10 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/src/jacobi-bugged.cpp:65]: Yes, thank you.

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 19]
```diff
@@ -1,177 +1,158 @@
...
+| Time to complete    | 1 hour for CPU or FPGA emulator; 2 hours for GPU
 
 ## Purpose
 
-The `jacobi` sample is a DPC++ application that solves a hardcoded 
-linear system of equations `Ax=b` using the Jacobi iteration. The matrix `A` 
-is `n x n` sparse matrix with diagonals [1 1 4 1 1], and vector `b` has
-a format `[6 7 8 ... 8 7 6]^T`. Thus, the solution vector `x` has a form of `[1 1 ... 1]^T`.
+The `jacobi` sample is a DPC++ application that solves a hardcoded linear system
+of equations `Ax=b` using the Jacobi iteration. The matrix `A` is `n x n` sparse
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/README.md:26]: Minor fix: is **an** `n x n` sparse matrix

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 20]
```diff
@@ -1,177 +1,158 @@
...
+matrix with diagonals `[1 1 4 1 1]`. Vectors `b` and `x` are `n x 1` vectors.
+Vector `b` is set in the way that all components of the solution vector `x` are 1.
 
-The program `jacobi-fixed` computes vector `x` correctly, while `jacobi-bugged`
-returns an incorrect result. With the debugger, you can put a breakpoint in and 
-outside of the kernel, step through the code, and check the values of local variables. 
-Use these features to find places in `jacobi-bugged` where the program does not 
-behave as it was intended.
+The program `jacobi-fixed` computes vector `x` correctly, that is, all
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/README.md:30]: Minor fix: ...correctly. That is, ...

File: Tools/ApplicationDebugger/jacobi/README.md [hunk 21]
```diff
@@ -1,177 +1,158 @@
...
+devices. For convenience, the `jacobi` code sample provides the ability to
+select the target device by using a command-line argument of `cpu`, `gpu`, or
+`accelerator`.
 
-The selected device is displayed in the output.
-For the overview of the debugger, please refer to 
-[Get Started Guide](https://software.intel.com/en-us/get-started-with-debugging-dpcpp)
-of the application debugger.
+The selected device is displayed in the output. For the overview of the
+debugger, please refer to [Get Started
```
Comment (barisaktemur) [inline Tools/ApplicationDebugger/jacobi/README.md:46]: Minor fix: ... refer to **the** Get Started...

**Review by barisaktemur — COMMENTED** [review]
A few minor comments. Overall it looks great to me. Thanks for this sample.

**Review by barisaktemur — APPROVED** [review]

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 1 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md:1]: @JoeOster Should I squash them into my original commit?

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 3 above)
Comment (JoeOster) [reply Tools/ApplicationDebugger/jacobi/README.md]: Then if it is unusable, we should have specific instruction or point to a set on how to get to an interactive prompt

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 1 above)
Comment (JoeOster) [reply Tools/ApplicationDebugger/jacobi/README.md:1]: they should already be in your commit

File: Tools/ApplicationDebugger/jacobi/third-party-programs.txt (hunk 11 above)
Comment (JoeOster) [reply Tools/ApplicationDebugger/jacobi/third-party-programs.txt:1]: It is not required unless your using third party components

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 3 above)
Comment (barisaktemur) [reply Tools/ApplicationDebugger/jacobi/README.md]: The command given right below (`$ qsub -I -l nodes=1:gpu:ppn=2`) gives the interactive prompt. Does it not suffice?

Comment (JoeOster) [reply Tools/ApplicationDebugger/jacobi/README.md]: Sorry, I don't use devcloud, so wasn't aware that the command did that. Regardless as this is a different mode than the batch mode we might want to state specifically... This command will get you into an interactive mode, allowing you to....

I am good if you don't want to point it out

**Comment by JoeOster** [comment]
@nsaiapova is this sample ready for review and merging? Friday, May 21st is the last day we can accept new samples for 2021.3. If so remove the WIP tag

File: Tools/ApplicationDebugger/jacobi/third-party-programs.txt [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+
```
Comment (JoeOster) [inline Tools/ApplicationDebugger/jacobi/third-party-programs.txt:2]: This file is not needed as your not including third-party components

**Review by JoeOster — CHANGES_REQUESTED** [review]

File: Tools/ApplicationDebugger/jacobi/third-party-programs.txt (hunk 11 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/third-party-programs.txt]: Removed.

File: Tools/ApplicationDebugger/jacobi/README.md (hunk 3 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/README.md]: I rephrased the section a bit.

**Review by nsaiapova — COMMENTED** [review]
> @nsaiapova is this sample ready for review and merging? Friday, May 21st is the last day we can accept new samples for 2021.3. If so remove the WIP tag

Sorry,  I thought that until all checklist items are checked the PR should be WIP.

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+      });
+      // Bug 3: the host buffers are not yet updated at this moment.
+      // Note, with this code, you get correct results on CPU.
+      // q.wait_and_throw();
+      // for (int i = 0; i < n; i++)
+      //   x_k[i] = x_k1[i];
+
+      // Fix bug 3.
+      q.submit([&](auto &h) {
+        auto acc_x_k = buffer_x_k.get_access<access::mode::write>(h);
```
Comment (pmpeter1) [inline Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp:94]: Should be using SYCL 2020 style accessors.

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Create 1D buffers for b vector and iteration vectors.
+    buffer buffer_b(b, range{n});
+    buffer buffer_x_k(x_k, range{n});
+    buffer buffer_x_k1(x_k1, range{n});
+
+    for (int k = 0; k < 15000; k++) {
+      // k-th iteration of Jacobi.
+      q.submit([&](auto &h) {
+        auto acc_b = buffer_b.get_access<access::mode::read>(h);
```
Comment (pmpeter1) [inline Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp:56]: Should be using SYCL 2020 style accessors.

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+         << q.get_device().get_info<info::device::name>() << "] from ["
+         << q.get_device().get_platform().get_info<info::platform::name>()
+         << "]\n";
+
+    // Create 1D buffers for b vector and iteration vectors.
+    buffer buffer_b(b, range{n});
+    buffer buffer_x_k(x_k, range{n});
+    buffer buffer_x_k1(x_k1, range{n});
+
+    for (int k = 0; k < 15000; k++) {
```
Comment (pmpeter1) [inline Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp:53]: 1500 is an arbitrary constant which would benefit from using a named constexpr.    Also for debugging perspective do you need this many iterations, or would something much smaller like 10 be enough to demo this?

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp (hunk 25 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp]: With my original matrix the convergence rate was indeed too low and required ~15000 iterations, as the matrix was not strictly diagonally dominant. I changed the matrix to [1 1 5 1 1], now the diagonal dominates (5 > 1+1+1+1) and the method converges in 100 iterations.
Also made the number of iterations a constexpr.
Thank you for this comment!

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp (hunk 23 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp:97]: Changed to `access_mode::write`, is that what you meant?

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp (hunk 24 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp]: Changed to `access_mode::read`.

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp (hunk 23 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp]: Now changed to `accessor acc_x_k (buffer_x_k, h, write_only);`

File: Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp (hunk 24 above)
Comment (nsaiapova) [reply Tools/ApplicationDebugger/jacobi/src/jacobi-fixed.cpp]: Now changed to accessor acc_b (buffer_b, h, read_only);

**Review by JoeOster — APPROVED** [review]
approved for ci only

**Review by pmpeter1 — APPROVED** [review]
